regtest test framework

Oscar-Pepper opened the pull request.

juanky201271 reviewed (commented):
I don't understand everything, of course... but if you change the `android` Cargo.toml... maybe can be a good idea to change the `ios` Cargo.toml... this make sense?

zancas commented on android/app/src/androidTest/java/org/ZingoLabs/Zingo/CommandsOnLaunch.kt:
```diff
@@ -29,51 +29,51 @@ class CommandsOnLaunch {
-        val server = "https://mainnet.lightwalletd.com:9067"
-        val chainhint = "main"
+        val server = "https://10.0.2.2:20000"
+        val chainhint = "regtest"
```
It'd be nice if we had 1-unified source for test data like this, that all our test libraries would load from!

Oscar-Pepper replied:
yeah i will add to kanban!

Oscar-Pepper commented:
ok so i have added a cargo nextest framework around the kotlin integration tests which we can run with `cargo nextest run --test-threads=1`:

- updated the current scripts to run specific tests or test suites
- updated all `init wallet` functions to include a monitor mempool option
- wrote a sync test which connects to a regtest server
- wrote cargo tests which setup a regtest server if that test needs it and calls the script

this cargo nextest should replace the use of `./scripts/integration_tests.sh`

IOS is not tested as i dont have the capability so please test and review thoroughly!

juanky201271 commented on android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt:
```diff
@@ -51,18 +79,22 @@ class CommandsOnLaunch {
     fun executeAddressFromUfvk() {
         val mapper = jacksonObjectMapper()
 
         val server = "https://mainnet.lightwalletd.com:9067"
```
Here the server could be `http://10.0.2.2:20000` ????

juanky201271 replied:
That's stupid of me... sorry.

Oscar-Pepper replied:
this is exactly what i have now implemented so not stupid! this makes the `offline` tests actually offline as in these tests cargo does not setup regtest and its pointed at main net for entropy bug tests etc.

zancas commented on android/app/src/androidTest/java/org/ZingoLabs/Zingo/OfflineTest.kt:
```diff
@@ -1,4 +1,4 @@
 package org.ZingoLabs.Zingo
 
-interface IntegrationTest {
+interface OfflineTest {
```
Because no traffic is leaving the local system?

Oscar-Pepper replied:
this is for tests that dont connect to a server. i.e. dont connect to regtest or do anything that requires syncing etc. although now i think about it this does imply that the wallet CANNOT connect to a server but that is not the case.... i will change so the tests dont get a valid mainnet server uri

zancas commented on android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt:
```diff
@@ -77,3 +109,57 @@ class CommandsOnLaunch {
+class ExecuteSyncFromSeed {
+    @Test
+    fun executeSyncFromSeed() {
+        val mapper = jacksonObjectMapper()
```
`jacksonObjectMapper`  ?    Interesting name.

Oscar-Pepper replied:
this was the one i chose to parse json ... there is a few options but this looked straight forward to use to me

zancas commented on android/app/src/main/java/org/ZingoLabs/Zingo/RPCModule.kt:
```diff
@@ -58,7 +58,7 @@ class RPCModule internal constructor(private val reactContext: ReactApplicationC
 
         // Create a seed
-        val seed = RustFFI.initnew(server, reactContext.applicationContext.filesDir.absolutePath, chainhint)
+        val seed = RustFFI.initnew(server, reactContext.applicationContext.filesDir.absolutePath, chainhint, "true")
```
This needs to be reviewed by @juanky201271 and/or @fluidvanadium , and tested with e2e/manually...    I find it curious that the type of the mempool monitor state selector passed to `initnew` seems to be a String instead of a bool.   I will review the `RustFFI.initnew` definition to understand the available types in that context.

Oscar-Pepper replied:
thanks i will implement a bool to the FFI

Oscar-Pepper replied:
i will implement this in a new PR as it is a bit more involved than i expected

zancas commented on rust/Dockerfile:
```diff
@@ -8,6 +8,7 @@ COPY lib/ /opt/zingo/rust/lib/
 COPY Cargo.lock /opt/zingo/rust/Cargo.lock
 COPY Cargo.toml /opt/zingo/rust/Cargo.toml
 COPY ios/ /opt/zingo/rust/ios/
+COPY test_utils/ /opt/zingo/rust/test_utils/
```
If this becomes too slow, we can consider mounting volumes from the build context, or publishing `test_utils` as a crate and making it a dependency.

Oscar-Pepper replied:
sure, copying the files is a very quick step in the build so unlikely to botteneck here but open to optimisations later down the line

zancas commented on rust/android/src/lib.rs:
```diff
@@ -33,13 +33,8 @@ pub unsafe extern "C" fn Java_org_ZingoLabs_Zingo_RustFFI_00024Companion_initnew
     j_serveruri: JString,
     j_data_dir: JString,
     j_chain_hint: JString,
+    j_monitor_mempool: JString,
```
See commentary elsewhere about preferring more informative types.

zancas commented on rust/android/src/lib.rs:
```diff
@@ -50,7 +45,21 @@ pub unsafe extern "C" fn Java_org_ZingoLabs_Zingo_RustFFI_00024Companion_initnew
         .unwrap();
 
-    let seed = rustlib::init_new(server_uri, data_dir, chain_hint);
+    let chain_hint = CString::from(CStr::from_ptr(
```
I believe `CStr::from_ptr` is unsafe.   Can we add assertions at this point to constrain possible values from this call?

One possibility is to mark the specific lines of this code which are unsafe, instead of the whole fn?

zancas replied:
I believe it's better to make the code marked `unsafe` be as specific as possible.   I think we can-and-should iteratively improve this.

Oscar-Pepper replied:
great comment. i think we should open a new PR for this after merge though as it applies to a lot of existing code and is a seperate matter

zancas commented on rust/android/tests/integration_tests.rs:
```diff
@@ -0,0 +1,34 @@
+
+    assert_eq!(exit_code, 0);
+
+    drop(child_process_handler);
```
Let's compare performance with and without this explicit `drop`..  not we'll _need_ a cph to be in scope..  so the code should have something like `_cph` or `_child_process_handler`, in the "implicit drop" version.

Oscar-Pepper replied:
its 45 seconds faster without drop on my test

Oscar-Pepper replied:
i have removed the explicit drop

zancas commented on rust/ios/src/lib.rs:
```diff
@@ -139,6 +134,40 @@ pub extern "C" fn initfromufvk(
     }
     .to_string();
+
+    let c_str = unsafe { CStr::from_ptr(monitor_mempool) };
```
This `unsafe` block seems to be a reasonable size, to me.  That having been said, I do not believe that `monitor_mempool` has the best type here.

zancas commented on rust/test_utils/Cargo.toml:
```diff
@@ -0,0 +1,7 @@
+edition = "2021"
+
+[dependencies]
+run_script = "^0.10.1"
```
I am skeptical that this crate is providing sufficient functionality to warrant a fresh dependency.   How about implementing a version that depends only on `std`..  and comparing the two implementations?

Oscar-Pepper replied:
i will change for std::process::Command

zancas reviewed (changes requested):
Overall this looks awesome.   The only review comment I consider blocking is the type of `monitor_mempool`.

juanky201271 reviewed (changes requested):
you need to include the new parameter `mempool` in a couple of places:
* https://github.com/Oscar-Pepper/zingo-mobile/blob/019fe6590055abb8286272d55f6fa873fe6cfeb2/ios/RPCModule.m#L239
* https://github.com/Oscar-Pepper/zingo-mobile/blob/019fe6590055abb8286272d55f6fa873fe6cfeb2/ios/RPCModule.m#L333

juanky201271 commented on ios/RPCModule.m:
```diff
@@ -247,6 +248,7 @@ -(NSString*) createNewWallet:(NSString*)server
                  birthday:(NSString*)birthday
                  server:(NSString*)server
                  chainhint:(NSString*)chainhint
+                 monitorMempool:(NSString*)monitorMempool
```
You need to follow the same pattern that you used in `android`, remove this line & put `true` directly in the call to `initfromseed`, like this: https://github.com/Oscar-Pepper/zingo-mobile/blob/019fe6590055abb8286272d55f6fa873fe6cfeb2/android/app/src/main/java/org/ZingoLabs/Zingo/RPCModule.kt#L77

Oscar-Pepper replied:
of course! thanks

juanky201271 commented on ios/RPCModule.m:
```diff
@@ -276,6 +278,7 @@ -(NSString*) createNewWallet:(NSString*)server
                  birthday:(NSString*)birthday
                  server:(NSString*)server
                  chainhint:(NSString*)chainhint
+                 monitorMempool:(NSString*)monitorMempool
```
You need to follow the same pattern that you used in `android`, remove this line & put `true` directly in the call to `initfromufvk`, like this: https://github.com/Oscar-Pepper/zingo-mobile/blob/019fe6590055abb8286272d55f6fa873fe6cfeb2/android/app/src/main/java/org/ZingoLabs/Zingo/RPCModule.kt#L93

zancas reviewed (approved):
Please open the `bool` for `monitor-mempool`!  And shrinking the unsafe blocks!

zancas commented:
OK... I went through and resolved all the conversations that I thought should have been resolved.

juanky201271 commented:
> you need to include the new parameter `mempool` in a couple of places:
> 
>     * https://github.com/Oscar-Pepper/zingo-mobile/blob/019fe6590055abb8286272d55f6fa873fe6cfeb2/ios/RPCModule.m#L239
> 
>     * https://github.com/Oscar-Pepper/zingo-mobile/blob/019fe6590055abb8286272d55f6fa873fe6cfeb2/ios/RPCModule.m#L333

My mistake, if you are using `true` directly in the calls, this comment make not sense, sorry.
